### PR TITLE
feat(linter): support lint vue files

### LINKED ISSUE #1238: feat(linter): Support lint the `<script>` part of Vue file
Since we have VSCode plugin and more and more rules, is it time to support lint the `<script>` part of Vue file?

### PR DESCRIPTION
**Pull request by mysteryven**
closes: #1238

### CODE REVIEW

**Comment by mysteryven** [comment]
Current dependencies on/for this PR:
* `main`
  * **PR #1706** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/1706?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/oxc-project/oxc/1706?utm_source=stack-comment).

File: crates/oxc_span/src/source_type.rs [hunk 1]
```diff
@@ -166,25 +174,222 @@ impl SourceType {
...
 
-        Ok(Self { language, module_kind: ModuleKind::Module, variant, always_strict: false })
+        Ok(Self { language, module_kind: ModuleKind::Module, variant, always_strict: false, range })
+    }
+    pub fn range(&self) -> Option<Span> {
+        self.range
+    }
+    fn may_parser_special_extension<P: AsRef<Path>>(path: P) -> (Option<bool>, Option<Span>) {
+        let default_ret = (None, None);
+        let Ok(content) = std::fs::read_to_string(path.as_ref()) else { return default_ret };
```
Comment (mysteryven) [inline crates/oxc_span/src/source_type.rs:213]: I don't really like here, it's means we need to read the content of Vue file twice. One here, another is in `process_path`
https://github.com/oxc-project/oxc/blob/23227305327b4a731095e3ac9641c3a63ee121d4/crates/oxc_linter/src/service.rs#L142-L153

How should I can improve it 😿

File: crates/oxc_span/src/source_type.rs [hunk 2]
```diff
@@ -166,25 +174,316 @@ impl SourceType {
...
+
+        let mut parser =
+            SpecialLanguageVariantParser::new(Path::new("test.vue"), source_text).unwrap();
+        parser.parse();
+        assert!(parser.is_ts);
+        assert_eq!(
+            get_js_content(&parser),
+            "
+                1/1
+            "
```
Comment (mysteryven) [inline crates/oxc_span/src/source_type.rs:465]: TODO: need to pass this

File: crates/oxc_span/src/source_type.rs [hunk 3]
```diff
@@ -166,25 +174,316 @@ impl SourceType {
...
+        );
+    }
```
Comment (mysteryven) [inline crates/oxc_span/src/source_type.rs:467]: TODO: need to pass this. Currently doesn't know whether `/` is the start of Reg Expression or an operator.

**Comment by Boshen** [comment]
The architecture around this is a bit more involved I think. We can support `.astro` if this is done right.

Let's iterate on this instead of having a huge PR.

File: crates/oxc_span/src/source_type.rs (hunk 3 above)
Comment (mysteryven) [reply crates/oxc_span/src/source_type.rs:467]: Oh, we don't need to check '</script>' in  RegExpression, `/</script>/` is not a valid syntax..

**Comment by aparajita** [comment]
Eagerly awaiting this! 🙏

File: crates/oxc_span/src/source_type.rs [hunk 4]
```diff
@@ -166,25 +174,311 @@ impl SourceType {
...
+
+                    if self.consume("script") {
+                        let open_tag_start = self.offset();
+                        self.consume_to('>');
+                        let open_tag_end = self.offset();
+                        start = open_tag_end;
+                        let attributes_text =
+                            Span::new(open_tag_start, open_tag_end).source_text(self.source_text);
+                        is_ts = attributes_text.contains(r#"lang="ts""#)
+                            || attributes_text.contains("lang='ts'");
```
Comment (magic-akari) [inline crates/oxc_span/src/source_type.rs:285]: I'm not very familiar with Vue, but I believe there might be another case here:

```html
<script lang=ts>
```

File: crates/oxc_span/src/source_type.rs [hunk 5]
```diff
@@ -166,25 +174,311 @@ impl SourceType {
...
+        let (is_ts_in_special_file, range) = Self::may_parser_special_extension(&path);
+
         let is_definition_file = file_name.ends_with(".d.ts")
             || file_name.ends_with(".d.mts")
             || file_name.ends_with(".d.cts");
 
         let language = match extension {
             "js" | "mjs" | "cjs" | "jsx" => Language::JavaScript,
             "ts" | "mts" | "cts" | "tsx" => Language::TypeScript { is_definition_file },
+            "vue" => {
```
[the remark this answers is not in the document]
Comment (mysteryven) [reply crates/oxc_span/src/source_type.rs:190]: Thank you, I am planing to support svelte in another way.

**Comment by mysteryven** [comment]
Will continue work on #1814

**Comment by Boshen** [comment]
vue support has been merged as a stack, see https://github.com/oxc-project/oxc/pull/1863#issuecomment-1872662826

Feel free to join, test and improve it together.

**Pull request closed without merging**